emails/management/commands/process_emails_from_sqs.py

```diff
@@ -398,68 +408,77 @@ def process_message(self, message):
-                    logger.info(f"processed sqs message ID: {message.message_id}")
-                except ClientError as e_retry:
-                    incr_if_enabled("message_from_sqs_error", 1)
-                    logger.error("sqs_client_error", extra=e_retry.response["Error"])
-                    results.update(
-                        {
-                            "success": False,
-                            "error": e_retry.response["Error"],
-                            "client_error_code": lower_error_code,
-                        }
-                    )
+        def success_callback(result: HttpResponse) -> None:
+            """Handle return from successful call to _sns_inbound_logic"""
+            # TODO: extract data from _sns_inbound_logic return
+            pass
+
+        def error_callback(exc_info: BaseException) -> None:
+            """Handle exception raised by _sns_inbound_logic"""
+            capture_exception(exc_info)
+            results["success"] = False
+            if isinstance(exc_info, ClientError):
+                incr_if_enabled("message_from_sqs_error")
+                err = exc_info.response["Error"]
+                logger.error("sqs_client_error", extra=err)
+                results["error"] = err
+                results["client_error_code"] = err["Code"].lower()
             else:
-                logger.error("sqs_client_error", extra=e.response["Error"])
-                results.update(
-                    {
-                        "success": False,
-                        "error": e.response["Error"],
-                        "client_error_code": lower_error_code,
-                    }
-                )
+                incr_if_enabled("email_processing_failure")
+                results["error"] = str(exc_info)
+                results["error_type"] = type(exc_info).__name__
+
+        # Run in a multiprocessing Pool
+        # This will start a subprocess, which needs to run django.setup
+        # The benefit is that the subprocess can be terminated
+        # The penalty is that is is slower to start
+        pool_start_time = time.monotonic()
+        with Pool(1, initializer=setup) as pool:
+            future = pool.apply_async(
+                _sns_inbound_logic,
+                [topic_arn, message_type, verified_json_body],
+                callback=success_callback,
+                error_callback=error_callback,
+            )
+            setup_time = time.monotonic() - pool_start_time
+            results["subprocess_setup_time_s"] = round(setup_time, 3)
+
+            message_start_time = time.monotonic()
+            message_duration = 0.0
+            while message_duration < self.max_seconds_per_message:
+                self.write_healthcheck()
+                future.wait(1.0)
+                message_duration = time.monotonic() - message_start_time
+                if future.ready():
+                    break
```

Comment:
question (non-blocking): Maybe related to how I don't understand `future._timeouts` property ... why do `wait(1.0)` rather than `wait(self.max_seconds_per_message)`? Is it to call `write_healthcheck` every second?

Reply:
Yes, exactly. If we don't write the healthcheck, then k8s will kill the process, and we won't have a chance to delete the stalled email from S3. This is especially bad in the DLQ - we just keep processing the same stalled email until it expires in 3 days. With this, we'll at least process one every three minutes, about 500 per day.